GH-2056 more use of getStatement instead of filter

hmottestad opened the pull request:
GitHub issue resolved: #2056 <!-- add a Github issue number here, e.g #123. This line 
                              automatically closes the issue when the PR is merged --> 

Briefly describe the changes proposed in this PR:

* Change from Iterator to Iterable
* Use Iterable where possible in helper methods in Models class
* Use .getStatements where possible in the RDF4J code base

jeenbroekstra commented on core/model/src/main/java/org/eclipse/rdf4j/model/impl/DynamicModel.java:
```diff
@@ -281,12 +296,32 @@ public void clear() {
+				&& contexts.length == 1) {
+			Statement statement = SimpleValueFactory.getInstance()
+					.createStatement(subject, predicate, object, contexts[0]);
+			Statement foundStatement = statements.get(statement);
```
Ah. Using a HashMap to be able to match and retrieve the original statement, instead of the newly created copy. Clever.

jeenbroekstra commented on core/sail/base/src/main/java/org/eclipse/rdf4j/sail/base/DistinctModelReducingUnionIteration.java:
```diff
@@ -34,15 +34,15 @@
 
 	DistinctModelReducingUnionIteration(CloseableIteration<? extends Statement, SailException> iterator, Model model,
-			Function<Model, Model> filterable) {
+			Function<Model, Iterator<Statement>> filterable) {
```
Can we keep the old constructor signature as a (deprecated) overload? Just to keep things backward compatible.

jeenbroekstra replied:
Actually never mind, I overlooked that this constructor is package-private anyway.

jeenbroekstra commented:
Btw are you aware that instead of labeling something WIP you can actually create a Draft pull request as well? 

![Screenshot from 2020-04-02 09-48-58](https://user-images.githubusercontent.com/728776/78193844-35e90900-74c7-11ea-85b6-8047d00680d2.png)

I'm fine with using WIP label too btw - as long as the intent is clear.

hmottestad commented:
I know. I did. But realized that it had the wrong target branch. Changed the target branch and then clicked the create button, but it had changed from draft to normal pr because I had changed the target branch. 

Unless you know how to make a branch draft after it’s created....

hmottestad commented:
Added benchmark is down from 19.2ms to 16.8ms, ~10% decrease. Not massive, but nice enough.

hmottestad commented on core/sail/base/src/main/java/org/eclipse/rdf4j/sail/base/SailDatasetImpl.java:
```diff
@@ -267,15 +267,16 @@ public void handleClose() throws SailException {
 		Model deprecated = changes.getDeprecated();
 		if (deprecated != null && iter != null) {
-			iter = difference(iter, deprecated.filter(subj, pred, obj, contexts));
+			iter = difference(iter, deprecated);
```
I don't think this change makes a difference for the results. The set of statements that are in `iter` but not in `deprecated` is the same unless the pattern used to filter `deprecated` is different from the pattern used by `iter`.

The difference is for performance. Now later on in the FilterIteration we can work directly with the hashmap without having to upgrade to a full LinkedHashModel because of a complicated `filter` operation.

jeenbroekstra replied:
The test suite appears to agree with you!

jeenbroekstra commented:
> I know. I did. But realized that it had the wrong target branch. Changed the target branch and then clicked the create button, but it had changed from draft to normal pr because I had changed the target branch.
> 
> Unless you know how to make a branch draft after it’s created....

No afraid not, this is where it falls down - I think you'd have to close the PR and create a completely new one. Labeling WIP is probably easier in those cases.

hmottestad commented on core/model/src/main/java/org/eclipse/rdf4j/model/Model.java:
```diff
@@ -191,14 +191,14 @@ public default boolean remove(Resource subj, URI pred, Value obj, Resource... co
 	 */
-	public default Iterator<Statement> getStatements(Resource subject, IRI predicate, Value object,
+	public default Iterable<Statement> getStatements(Resource subject, IRI predicate, Value object,
 			Resource... contexts) {
-		return filter(subject, predicate, object, contexts).iterator();
+		return () -> filter(subject, predicate, object, contexts).iterator();
```
@jeenbroekstra I'm proposing we move from iterator to iterable because iterable supports for-each loops and is also simpler to convert into a stream.

hmottestad replied:
It's a pity that Java didn't introduce a combo interface with both iterable and stream. We can always create our own interface in the future though with extends both and still remain backwards compatible.

jeenbroekstra replied:
> @jeenbroekstra I'm proposing we move from iterator to iterable because iterable supports for-each loops and is also simpler to convert into a stream.

I'm fine with that, but can you make sure you update the original issue's (#2030) description, and also the javadoc in Model?

hmottestad commented:
I feel happy with this now and ready to merge.
